File: app/src/main/java/com/feriantes4dawin/feriavirtualmovil/ui/login/LoginActivity.java
package com.feriantes4dawin.feriavirtualmovil.ui.login;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.Editable;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.*;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.text.HtmlCompat;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;

import com.feriantes4dawin.feriavirtualmovil.FeriaVirtualApplication;
import com.feriantes4dawin.feriavirtualmovil.FeriaVirtualComponent;
import com.feriantes4dawin.feriavirtualmovil.R;
import com.feriantes4dawin.feriavirtualmovil.data.repos.UsuarioRepositoryImpl;
import com.feriantes4dawin.feriavirtualmovil.ui.main.MainActivity;
import com.feriantes4dawin.feriavirtualmovil.ui.util.FeriaVirtualConstants;
import com.feriantes4dawin.feriavirtualmovil.ui.util.SimpleTextWatcher;
import com.google.android.material.snackbar.Snackbar;
import com.google.gson.Gson;

import javax.inject.Inject;

public class LoginActivity extends AppCompatActivity {

    //private lateinit var loginViewModel: LoginViewModel

    private FeriaVirtualComponent feriaVirtualComponent;

    @Inject
    public UsuarioRepositoryImpl usuarioRepository;

    @Inject
    public Gson convertidorJSON;

    private LoginViewModel loginViewModel;
    private LoginViewModelFactory loginViewModelFactory;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_login);

        //Injectamos nuestras dependencias
        feriaVirtualComponent =  ((FeriaVirtualApplication) getApplicationContext())
                .getFeriaVirtualComponent();

        feriaVirtualComponent.injectUsuarioRepositoryIntoLoginActivity(this);

        //Creamos nuestro factory!
        loginViewModelFactory = new LoginViewModelFactory(
                usuarioRepository,
                convertidorJSON,
                (FeriaVirtualApplication) getApplicationContext());

        //Creamos el viewmodel con nuestro factory!
        loginViewModel = new ViewModelProvider(this,loginViewModelFactory)
            .get(LoginViewModel.class);

        //Obtengo las preferencias compartidas (por si acaso!)
        SharedPreferences sp = getSharedPreferences(
                FeriaVirtualConstants.FERIAVIRTUAL_MOVIL_SHARED_PREFERENCES,
                Context.MODE_PRIVATE);

        if(revisarSiHaySesion(sp)){

            manejarSesion();

        } else {

            EditText username = findViewById(R.id.username);
            EditText password = findViewById(R.id.password);
            Button login = findViewById(R.id.login);
            LinearLayout loading = findViewById(R.id.dcp_llloading);
            Button  quit = findViewById(R.id.btnQuit);

            TextView linkRegistro = findViewById(R.id.tvLinkRegistro);
            Spanned textolink;

            textolink = HtmlCompat.fromHtml(
                    String.format("%s%s%s",
                            "<a href='https://www.google.com/?hl=es'>",
                            getResources().getString(R.string.launch_user_registry_web),
                            "</a>"),
                    HtmlCompat.FROM_HTML_MODE_COMPACT);

            //Le da el aspecto y acciones de un enlace
            linkRegistro.setText(textolink);
            linkRegistro.setMovementMethod(LinkMovementMethod.getInstance());

            /*
            linkRegistro.setOnClickListener {

                val uri = "https://www.google.com/?hl=es"
                val intent = Intent(Intent.ACTION_VIEW)
                intent.data = Uri.parse(uri)
                startActivity(intent)

            }*/

            loginViewModel.loginFormState.observe(this, new Observer<LoginFormState>() {
                @Override
                public void onChanged(LoginFormState loginFormState) {

                    if(loginFormState == null){
                        return;
                    }

                    // disable login button unless both username / password is valid
                    login.setEnabled(loginFormState.isDataValid);

                    if (loginFormState.usernameError != null) {
                        username.setError(getString(R.string.invalid_username));
                    } else {
                        username.setError(null);
                    }
                    if (loginFormState.passwordError != null) {
                        password.setError(getString(R.string.invalid_password));
                    } else {
                        password.setError(null);
                    }
                }
            });

            loginViewModel.loginResult.observe(this, new Observer<LoginResult>() {
                @Override
                public void onChanged(LoginResult loginResult) {

                    if(loginResult == null){
                        return;
                    }

                    loading.setVisibility(View.GONE);

                    if (loginResult.error != null) {
                        showLoginFailed(loginResult.error);
                    }
                    if (loginResult.success != null) {

                        updateUiWithUser(loginResult.success);
                        manejarSesion();

                    }

                }
            });

            username.addTextChangedListener(
                    new SimpleTextWatcher(null) {
                        @Override
                        public void afterTextChanged(Editable s) {
                            loginViewModel.loginDataChanged(
                                    username.getText().toString(),
                                    password.getText().toString()
                            );
                        }
                    }
            );

            password.addTextChangedListener(
                    new SimpleTextWatcher(null) {
                        @Override
                        public void afterTextChanged(Editable s) {
                            loginViewModel.loginDataChanged(
                                    username.getText().toString(),
                                    password.getText().toString()
                            );
                        }
                    }
            );

            password.setOnEditorActionListener(new TextView.OnEditorActionListener() {
                @Override
                public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                    if(actionId == EditorInfo.IME_ACTION_DONE){
                        loginViewModel.login(
                                username.getText().toString(),
                                password.getText().toString()
                        );
                    }
                    return false;
                }
            });

            login.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    loading.setVisibility(View.VISIBLE);
                    loginViewModel.login(username.getText().toString(), password.getText().toString());
                }
            });

            quit.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    finish();
                }
            });

        }

    }

    private void updateUiWithUser(LoggedInUserView model) {
        String welcome = getString(R.string.welcome);
        String displayName = model.displayName;
        // TODO : initiate successful logged in experience
        Toast.makeText(
            getApplicationContext(),
            String.format("%s, %s",welcome,model.displayName),
            Toast.LENGTH_LONG
        ).show();
    }

    private void showLoginFailed(int errorString) {
        View v = findViewById(R.id.actlog_svPrincipal);
        Snackbar.make(v, getString(errorString), Snackbar.LENGTH_SHORT).show();
    }


    public boolean revisarSiHaySesion(SharedPreferences sp){

        return (!sp.getString(FeriaVirtualConstants.SP_FERIAVIRTUAL_WEBAPI_AUTH_TOKEN,"").equals(""))
                && (!sp.getString(FeriaVirtualConstants.SP_USUARIO_OBJ_STR,"").equals(""));

    }


    public void manejarSesion(){

        //Cambiamos a la actividad principal
        Intent menuprincipalintent = new Intent(getApplicationContext(), MainActivity.class);

        startActivity(menuprincipalintent);

        setResult(Activity.RESULT_OK);

        //Complete and destroy login activity once successful
        finish();

    }
}

File: app/src/main/java/com/feriantes4dawin/feriavirtual/ui/login/LoginActivity.kt
package com.feriantes4dawin.feriavirtual.ui.login

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.Spanned
import android.text.TextWatcher
import android.text.method.LinkMovementMethod
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.core.text.HtmlCompat.TO_HTML_PARAGRAPH_LINES_INDIVIDUAL
import androidx.core.text.HtmlCompat.fromHtml
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.feriantes4dawin.feriavirtual.R
import com.feriantes4dawin.feriavirtual.ui.main.MainActivity

class LoginActivity : AppCompatActivity() {

    //private lateinit var loginViewModel: LoginViewModel
    private lateinit var loginViewModel : LoginViewModel
    private var loginViewModelFactory  = LoginViewModelFactory()

    @SuppressLint("WrongConstant")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_login)

        loginViewModel = ViewModelProvider(this,loginViewModelFactory)
            .get(LoginViewModel::class.java)

        val username = findViewById<EditText>(R.id.username)
        val password = findViewById<EditText>(R.id.password)
        val login = findViewById<Button>(R.id.login)
        val loading = findViewById<ProgressBar>(R.id.loading)
        val quit = findViewById<Button>(R.id.btnQuit)

        val linkRegistro = findViewById<TextView>(R.id.tvLinkRegistro)
        lateinit var textolink:Spanned

        textolink = fromHtml(
        "<a href='https://www.google.com/?hl=es'>${resources.getString(R.string.launch_user_registry_web)}</a>",
            TO_HTML_PARAGRAPH_LINES_INDIVIDUAL)

        //Le da el aspecto y acciones de un enlace
        linkRegistro.text = textolink
        linkRegistro.movementMethod = LinkMovementMethod.getInstance()

        /*
        linkRegistro.setOnClickListener {

            val uri = "https://www.google.com/?hl=es"
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = Uri.parse(uri)
            startActivity(intent)

        }*/

        loginViewModel.loginFormState.observe(this@LoginActivity, Observer {
            val loginState = it ?: return@Observer

            // disable login button unless both username / password is valid
            login.isEnabled = loginState.isDataValid

            if (loginState.usernameError != null) {
                username.error = getString(loginState.usernameError)
            }
            if (loginState.passwordError != null) {
                password.error = getString(loginState.passwordError)
            }
        })

        loginViewModel.loginResult.observe(this@LoginActivity, Observer {
            val loginResult = it ?: return@Observer

            loading.visibility = View.GONE
            if (loginResult.error != null) {
                showLoginFailed(loginResult.error)
            }
            if (loginResult.success != null) {

                updateUiWithUser(loginResult.success)

                //Cambiamos a la actividad principal
                val menuprincipalintent = Intent(this, MainActivity::class.java)

                startActivity(menuprincipalintent)
            }
            setResult(Activity.RESULT_OK)

            //Complete and destroy login activity once successful
            finish()
        })

        username.afterTextChanged {
            loginViewModel.loginDataChanged(
                username.text.toString(),
                password.text.toString()
            )
        }

        password.apply {
            afterTextChanged {
                loginViewModel.loginDataChanged(
                    username.text.toString(),
                    password.text.toString()
                )
            }

            setOnEditorActionListener { _, actionId, _ ->
                when (actionId) {
                    EditorInfo.IME_ACTION_DONE ->
                        loginViewModel.login(
                            username.text.toString(),
                            password.text.toString()
                        )
                }
                false
            }

            login.setOnClickListener {
                loading.visibility = View.VISIBLE
                loginViewModel.login(username.text.toString(), password.text.toString())
            }
        }

        quit.setOnClickListener {
            finish()
        }
    }

    private fun updateUiWithUser(model: LoggedInUserView) {
        val welcome = getString(R.string.welcome)
        val displayName = model.displayName
        // TODO : initiate successful logged in experience
        Toast.makeText(
            applicationContext,
            "$welcome $displayName",
            Toast.LENGTH_LONG
        ).show()
    }

    private fun showLoginFailed(@StringRes errorString: Int) {
        Toast.makeText(applicationContext, errorString, Toast.LENGTH_SHORT).show()
    }
}

/**
 * Extension function to simplify setting an afterTextChanged action to EditText components.
 */
fun EditText.afterTextChanged(afterTextChanged: (String) -> Unit) {
    this.addTextChangedListener(object : TextWatcher {
        override fun afterTextChanged(editable: Editable?) {
            afterTextChanged.invoke(editable.toString())
        }

        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
    })
}